### PR TITLE
👌 Improve footnote def/ref warnings and translations

### LINKED ISSUE #445: Footnotes translation
### Describe the problem

The expected behaviour of footnotes with text before and after when they are translated is wrong  (without text, everything seems to be ok).

The reference to footnotes like `[^2]` doesn't work. The only way that I can reference to footnotes is `[^2]: `, but this way works only if the footnotes doesn't have text before and after.

It fails with:
`footnotes_i18n/source/introduction.md:8: WARNING: inconsistent footnote references in translated message. original: ['[^2]'], translated: []`

### Link to your repository or website

https://github.com/ValentinConstantin/footnotes_i18n

### Steps to reproduce

python3 -m venv .venv
source .venv/bin/activate
pip install pip wheel setuptools
pip install -r requirements.txt
make -e SPHINXOPTS="-v -D language='fr'" html
xdg-open build/html/index.html
make -e SPHINXOPTS="-v -D language='en'" html

### The version of Python you're using

Python 3.9.8

### Your operating system

_No response_

### Versions of your packages

_No response_

### Additional context

_No response_

### LINKED ISSUE #801: Footnotes not rendering as expected in localized version
### What version of `myst-parser` are you using?

2.0.0

### What version dependencies are you using?
sphinx-build (7.2.6)
sphinx-rtd-theme
sphinx-design
sphinxcontrib-mermaid
sphinxcontrib-svg2pdfconverter
sphinx-intl

### What operating system are you using?

Linux

### Describe the Bug

I have a Sphinx documentation using *myst* and *sphinx-intl*. In the default English version of the documentation, footnotes are formatted correctly, and I'm getting properly rendered footnotes in the resulting HTML when I build the documentation.

However, I've also created a localized German version of the documentation, and while the translation seems to work fine overall, I'm encountering an issue with footnotes. In the German version, the footnotes are not being converted to the correct HTML format. Instead of seeing the formatted footnotes as in the English version, it appears that the German version is simply printing the Markdown string, like `[^1]`, without converting it to the expected HTML footnote markup. The footnote section works fine, but the inline link is not formatted and linked correctly.

```sh
/docs/source/test.md:1: WARNING: inconsistent footnote references in translated message. original: ['[^1]'], translated: []                                                                                                                   
```

test

The .po files look fine, e.g.:

```
#: ../test.md:1
msgid ""
"Test[^1]"
msgstr ""
"Test[^1]"
```

### Expected Behavior

I expect that when I build the German version of the documentation, the footnotes should be processed and rendered in the same way as they are in the original English version. Other markdown features also work without interfering with the locales.

### To Reproduce

_No response_

### LINKED ISSUE #884: Message "inconsistent footnote references in translated message...."
### What version of `myst-parser` are you using?

2.0.0

### What version dependencies are you using?

Docutils 0.19, Python 3.11.1, on win32
sphinx-7.2.6

### What operating system are you using?

Windows

### Describe the Bug

In translated documents footnotes does not reprouced and this error appears:
**`WARNING: inconsistent footnote references in translated message. original: ['[^anothernote]'], translated: []`**

[Sample.md](https://github.com/executablebooks/MyST-Parser/files/14526678/Sample.md)

### Expected Behavior

footnotes must be displayed in translated documents too!

#### Setup a fresh document
```
mkdir test_02 & cd test_02
sphinx-quickstart --sep -p TestBug -a BravoBaldo -r "0" -l "en" --extensions "myst_parser"
echo locale_dirs = ['locale/']            >> source\conf.py
```
##### Create and include a simple file called "Sample.md" with some footnotes (see [attached Sample.md](https://github.com/executablebooks/MyST-Parser/files/14526678/Sample.md)).
`make singlehtml`
**Result: all ok.**

### Internationalization:
```
cls & make clean & make gettext
sphinx-intl update -p build/gettext -l en -l it

make clean & sphinx-build -b singlehtml -D language=it ./source build/singlehtml/it
```
**Result: Still ok.**

##### Translate "source\locale\it\LC_MESSAGES\Sample.po" (**is enough a little change**)

> Note: this means to fill a "msgstr" with a translation of matching "msgid" for example:
> `msgid "Here's a simple footnote,[^1] and here's a longer one.[^bignote]"`
> `msgstr "XXXHere's a simple footnote,[^1] and here's a longer one.[^bignote]"`
> (simply add "XXX")
> _(Strangely, if the text is the same the bug does not occur!!_)

`make clean & sphinx-build -b singlehtml -D language=it ./source build/singlehtml/it`

This error appears for each note:
**`WARNING: inconsistent footnote references in translated message. original: ['[^anothernote]'], translated: []`**
and the footnote(s) are not displayed

### LINKED ISSUE #930: Add warning for invalid footnotes reference (and unused definitions)
### What version of `myst-parser` are you using?

3.0.1

### What version dependencies are you using?

Sphinx 7.2.6
docutils 0.20.1

### What operating system are you using?

Mac

### Describe the Bug

If a Markdown footnote does not have a corresponding reference in either direction, no warnings are emitted. Compare the equivalent RST and Markdown:

```rst
Lorem ipsum [#f1]_ dolor sit amet ... [#f2]_

.. rubric:: Footnotes

.. [#f1] Text of the first footnote.
.. [#f3] Text of the second footnote.
```

````md
Lorem ipsum [^f1] dolor sit amet ... [^f2]

```{rubric} Footnotes
```

[^f1]: Text of the first footnote.

[^f3]: Text of the second footnote.
````

The first one gives two warnings in the build

```
test.rst:1: ERROR: Too many autonumbered footnote references: only 0 corresponding footnotes available.
test.rst:1: ERROR: Unknown target name: "f2".
```

The second one gives no warnings. The resulting document just has a literal [^f2] in the text. 

### Expected Behavior

An invalid footnote should produce a warning, ideally in both directions, since it is an invalid cross-reference. 

### To Reproduce

_No response_

### PR DESCRIPTION
**Pull request by chrisjsewell**
Footnotes are now parsed similar to the corresponding restructuredtext, in that resolution (between definitions and references) and ordering is now deferred to transforms on the doctree.

This allows for the proper interaction with other docutils/sphinx transforms, including those that perform translations.

In addition, an upstream improvement to unreferenced footnote definitions is also added here: https://github.com/sphinx-doc/sphinx/pull/12730, so that unreferenced and duplicate definitions are correctly warned about, e.g.:

```
source/index.md:1: WARNING: Footnote [1] is not referenced. [ref.footnote]
source/index.md:4: WARNING: Duplicate footnote definition found for label: 'a' [ref.footnote]
```

It is of note that warnings for references with no corresponding definitions are deferred to docutils to handle, e.g. for `[^a]` with no definition:

```
source/index.md:1: ERROR: Too many autonumbered footnote references: only 0 corresponding footnotes available. [docutils]
source/index.md:1: ERROR: Unknown target name: "a". [docutils]
```

These warning messages are a little obscure, and it would be ideal that one clear warning was emitted for the issue.
However, it is non-trivial in this extension; to both suppress the existing warnings, and then replace them with a better one,
so for now we don't do it here, and ideally this would be improved upstream in docutils.

closes #930
closes #884
closes #801
closes #445

### CODE REVIEW

**Comment by chrisjsewell** [comment]
All good @asmeurer 😄 ?

**Comment by asmeurer** [comment]
I tried this but I got an exception:

```
Traceback (most recent call last):
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/cmd/build.py", line 298, in build_main
    app.build(args.force_all, args.filenames)
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/application.py", line 355, in build
    self.builder.build_update()
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 293, in build_update
    self.build(to_build,
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 313, in build
    updated_docnames = set(self.read())
                           ^^^^^^^^^^^
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 420, in read
    self._read_serial(docnames)
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 441, in _read_serial
    self.read_doc(docname)
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 498, in read_doc
    publisher.publish()
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/docutils/core.py", line 234, in publish
    self.document = self.reader.read(self.source, self.parser,
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/io.py", line 105, in read
    self.parse()
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/docutils/readers/__init__.py", line 76, in parse
    self.parser.parse(self.input, document)
  File "/Users/aaronmeurer/Documents/MyST-Parser/myst_parser/parsers/sphinx_.py", line 73, in parse
    parser = create_md_parser(config, SphinxRenderer)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/MyST-Parser/myst_parser/parsers/mdit.py", line 64, in create_md_parser
    .use(footnote_plugin, inline=False, move_to_end=False, always_match_refs=True)
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/markdown_it/main.py", line 253, in use
    plugin(self, *params, **options)
TypeError: footnote_plugin() got an unexpected keyword argument 'inline'
```

**Comment by chrisjsewell** [comment]
> I tried this but I got an exception:

you haven't updated your mdit-py-plugin version

**Comment by asmeurer** [comment]
I have the latest version from conda. Is there a dev version I need to use?

**Comment by chrisjsewell** [comment]
> I have the latest version from conda. Is there a dev version I need to use?

yeh it was only released on pypi today, so this has only just appeared https://github.com/conda-forge/mdit-py-plugins-feedstock/pull/21

**Comment by asmeurer** [comment]
OK, I was able to get it working. Two things I noticed:

- The line number for `WARNING: No footnote definition found for label:` is wrong. It is giving the line for the beginning of the paragraph instead of the line with the label.

- The error message `ERROR: Too many autonumbered footnote references: only 0 corresponding footnotes available.` is confusing. I'm not sure I understand what it's trying to say. And it seems to be a duplicate error for the same line that gives `ERROR: Unknown target name:`.

File: myst_parser/mdit_to_docutils/base.py [hunk 1]
```diff
@@ -279,30 +280,54 @@ def _render_finalise(self) -> None:
...
+                        f"No footnote definition found for label: '{foot_label}'",
+                        MystWarnings.MD_FOOTNOTE_MISSING,
+                        line=node.line,
+                        append_to=self.current_node,
+                    )
+                    # lets remove the footnote references, so that docutils does not produce any more warnings
+                    # we need to replace them with an element though, so that ids can be moved over (otherwise docutils excepts)
+                    if node.get("auto"):
+                        self.document.autofootnote_refs.remove(node)
+                    node.replace_self(nodes.inline(text=f"[^{node['refname']}]"))
```
Comment (chrisjsewell) [inline myst_parser/mdit_to_docutils/base.py:310]: code to suppress docutils warnings

File: tests/test_renderers/fixtures/reporter_warnings.md [hunk 2]
```diff
@@ -110,14 +110,28 @@ Non-consecutive headings:
 <string>:2: (WARNING/2) Non-consecutive header level increase; H1 to H3 [myst.header]
 .
 
-multiple footnote definitions
+footnote reference with no definition
+.
+[^a]
+.
+<string>:1: (WARNING/2) No footnote definition found for label: 'a' [myst.footnote]
+.
```
Comment (chrisjsewell) [inline tests/test_renderers/fixtures/reporter_warnings.md:118]: this produced additional docutils warnings, before the fix (https://github.com/executablebooks/MyST-Parser/pull/931/files#r1598103576)

**Comment by chrisjsewell** [comment]
> It is giving the line for the beginning of the paragraph instead of the line with the label.

ah well, if you haven't noticed this the same for every "inline" warning in sphinx/docutils (for both restructuredtext and myst)

> The error message `ERROR: Too many autonumbered footnote references: only 0 corresponding footnotes available.` is confusing, ...

these are all from docutils, not myst
but thats strange, because there is specific code to suppress these (see https://github.com/executablebooks/MyST-Parser/pull/931/files#r1598103576) and they don't show up in the tests anymore (https://github.com/executablebooks/MyST-Parser/pull/931/files#r1598109555)

do you have a minimal working example, that could get the test to produce these errors

**Comment by chrisjsewell** [comment]
I definitely get "good" warnings now, if I add some test examples to the top of `docs/index.md`

image

image

**Comment by asmeurer** [comment]
Sorry, don't have time to minimize it, but it's coming from https://github.com/Quansight-Labs/ndindex/pull/136, commit 09541959137694579060a6c48b3e429fa8e30026

**Comment by asmeurer** [comment]
It's possible it's coming from the other cross-reference in the same paragraph. I had to use a workaround to cross-reference cross-page footnotes. These errors all came from me splitting a document into multiple pages and forgetting to move the footnotes.

**Pull request merged**